docs/design/gluster-block-provisioning.md

```diff
@@ -0,0 +1,350 @@
+        } `json:"blockvolume"`
+}
+
+```
+
+
+#### BlockVolumeListResponse
+
+The block volume list request just gets a list
+of block volume IDs as response.
+
+
+```
+type BlockVolumeListResponse struct {
+        BlockVolumes []string `json:"blockvolumes"`
+}
+
+```
+
+### Details About Heketi's Internal Logic
+
+#### Block-hosting volumes
+
+The loopback files for block volumes need to be stored in
+gluster file volumes. Volumes used for gluster-block volumes
+should not be used for other purposes. For want of a better
+term, we call these volumes that can host block-volume
+loopback files **block-hosting file-volumes** or
+**block-hosting volumes** in this document.
+
+#### Labeling block-hosting volumes
+
+In order to satisfy a blockvolume create request, Heketi
+needs to find and appropriate block-hosting volume in
+the available clusters.  Hence heketi should internally
+flag these volumes with a label (`block`).
+
+#### Block-hosting volume creation automatism
+
+When heketi, upon receiving a blockvolume create request,
+does not find a block-hosting volume with sufficient
+space in any of the considered clusters, it would look for
+sufficient unused space in the considered clusters and create
+a new gluster file volume, or expand an existing volume
+labeled `block`.
+
+The sizes to be used for auto-creation of block-hosting
+volumes will be subject to certain parameters that can
+be configured and will have reasonable defaults:
+
+* `CreateBlockHostingVolumes`: Enable auto-creation of
+  block-hosting volumes?
+  Defaults to **yes**.
+* `ExpandBlockHostingVolumes`: Should heketi expand an
+  existing block-hosting volume if sufficient space is
+  available on the same cluster?
+  Defaults to **No**.
+* `BlockHostingVolumeNewSize`: The size for a new block-hosting
+  volume to be created on a cluster will be the minimum of the value
+  of this setting and maximum size of a volume that could be created.
+  This size will also be used when expanding volumes: The amount
+  added to the existing volume will be the minimum of this value
+  and the maximum size that could be added.
+  Defaults to **1TB**.
+
+#### Internal heketi db format for block volumes
+
+Heketi stores information about the block volumes
+in it's internal DB. The information stored is
+
+* id: id of this block volume
+* name: name given to the volume
+* volume: the id of the block-hosting volume where the loopback file resides
+* hosts: the target ips for this volume
+
+#### Cluster selection
+
+By default, heketi would consider all available clusters
+when looking for space to create a new block-volume file.
+
+With the clusters request parameter, this search can be
+narrowed down to an explicit list of one or more clusters.
+With the help of the ```clusterid``` storage class option,
+this gives the kubernetes administrator a way to e.g. separate
+different storage qualities, or to reserve a cluster
+exclusively for block-volumes.
+
```

Comment:
We thought about it. But we don't want to be that restrictive. The above comment explains how to achieve that separation by configuring different storage classes. It might be best practice to set it up that way. But the default should be the most simple solution where a single gluster cluster can host both regular GlusterFS PVs as well as block-hosting volumes with a lot of loopback devices.